torch/optim/adagrad.py

```diff
@@ -349,7 +349,12 @@ def _multi_tensor_adagrad(
                 device_state_sums[i] = torch.view_as_real(device_state_sums[i])
 
         # Update steps
-        torch._foreach_add_(device_state_steps, 1)
+        # If steps are on CPU, foreach will fall back to the slow path, which is a for-loop calling t.add(1) over
+        # and over. 1 will then be wrapped into a Tensor over and over again, which is slower than if we just
+        # wrapped it now and passed it to the API as a tensor already. So we wrap it now.
+        step_device = device_state_steps[0].device
+        one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
+        torch._foreach_add_(device_state_steps, one, 1.0)
```

Comment:
For some reason, specifying an alpha (it could be 1.0 or just 1) is necessary for the perf win. I have no idea why the default of 1 is different and I'm a bit tired to look into this now. This is certainly interesting and deserves figuring out though.

Reply:
Should we check if alpha has any regression if "step" is on GPU? (assuming that "step" not being on CPU generally implies it is on GPU)

Reply:
that’s a good idea, will do tomorrow

Reply:
I'm gonna use an if else blatantly. I've also discovered the reason why the alpha specification is necessary, and it does not seem easily fixable: https://github.com/pytorch/pytorch/issues/111157